src/smt/solver_engine.h

```diff
@@ -460,6 +461,38 @@ class CVC5_EXPORT SolverEngine
    * of type T.
    */
   void declarePool(const Node& p, const std::vector<Node>& initValue);
+
+  /**
+   * Add an oracle function to the state, also adds an oracle interface
+   * defining it, when binName is not "".
+   *
+   * @param var The oracle function
+   * @param binName The name of the binary, if one exists
+   */
+  void declareOracleFun(Node var, const std::string& binName = "");
```

Comment:
I wonder if it would be possible to generalize this interface a bit. I am thinking of something like:
```suggestion
  void declareOracleFun(Node var, std::function<std::string(const std::string&)> oracle);
```
or even better
```suggestion
  void declareOracleFun(Node var, std::function<Node(Node)> oracle);
```
With this [`std::function`](https://en.cppreference.com/w/cpp/utility/functional/function) interface, users (including us) can directly provide their oracle as a function, instead of a program. This interface has 3 potential benefits:
1. It's easier to use and allows us to provide examples in one file.
1. It is faster for cases where users have a C++ interface/implementation for their oracle.
3. Most importantly, with the `Node` version, I think we can also avoid having a `Node` level parser 😬.

Reply:
Yes, it would be interesting to give an interface like this.

The current interface is to support API users and users of the text interface.

Regardless, indeed it would be interesting to pursue extensions where the oracle has been "compiled" into the cvc5 code base. I would consider this to be a separate functionality, where the internal oracle utilities can be generalized later for this.